pkg/scheduler/internal/cache/node_tree.go

```diff
@@ -135,36 +121,32 @@ func (nt *nodeTree) updateNode(old, new *v1.Node) {
 	nt.addNode(new)
 }
 
-func (nt *nodeTree) resetExhausted() {
-	for _, na := range nt.tree {
-		na.lastIndex = 0
-	}
-	nt.zoneIndex = 0
-}
-
-// next returns the name of the next node. NodeTree iterates over zones and in each zone iterates
+// listNodes returns the list of namesof the node. NodeTree iterates over zones and in each zone iterates
 // over nodes in a round robin fashion.
-func (nt *nodeTree) next() string {
+func (nt *nodeTree) list() ([]string, error) {
+	nodesList := make([]string, 0, nt.numNodes)
 	if len(nt.zones) == 0 {
-		return ""
+		return nodesList, nil
 	}
 	numExhaustedZones := 0
-	for {
-		if nt.zoneIndex >= len(nt.zones) {
-			nt.zoneIndex = 0
-		}
-		zone := nt.zones[nt.zoneIndex]
-		nt.zoneIndex++
-		// We do not check the exhausted zones before calling next() on the zone. This ensures
-		// that if more nodes are added to a zone after it is exhausted, we iterate over the new nodes.
-		nodeName, exhausted := nt.tree[zone].next()
-		if exhausted {
-			numExhaustedZones++
-			if numExhaustedZones >= len(nt.zones) { // all zones are exhausted. we should reset.
-				nt.resetExhausted()
+	allocatedNodes := 0
+	nodeIndex := 0
+	for allocatedNodes < nt.numNodes {
+		for zoneIndex := 0; zoneIndex < len(nt.zones); zoneIndex++ {
```

Comment:
This doesn't sound correct to me, we need to start from the last used zone, not from zoneIndex=0.

Reply:
with this loop, we iterate over all the zones, adding a node from each non exhausted loop until we have gone through all the zones. then we go through the loop again. The change did not modify the output of the list() function. 

Or do you mean to say that a new run of list() should pick up from where the last list() run ended ?

Reply:
ah, ok, I missed that the append happens inside the zone loop.